Document recursive chown feature (#19391)
kubernetes/website: content/en/docs/tasks/configure-pod-container/security-context.md

-rw-r--r-- 1 1000 2000 6 Jun  6 20:08 testfile
```

> Kubernetes by default recursively changes ownership and permission of volume to match `fsGroup` specified in pod's

Reviewer: Add the feature header that this is an alpha feature in 1.18

Author: fixed

`securityContext` when volume is mounted. Changing ownership and permissions of a large volume can be time consuming
operation and that is where `fsGroupChangePolicy` field in `securityContext` comes in.

> **fSGroupChangePolicy** -  fSGroupChangePolicy defines behavior of changing ownership and permission of the volume

Reviewer: Should we document the two fields here?

Author: added

exit
```

## Configure volume permission and ownership change policy for Pods

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

Kubernetes by default recursively changes ownership and permission of volume to match `fsGroup` specified in pod's
`securityContext` when volume is mounted. Changing ownership and permissions of a large volume can be time consuming

> operation and that is where `fsGroupChangePolicy` field in `securityContext` comes in.

Reviewer 2: nit / my suggestion (only revise if something bigger also needs fixing, or if you want to):
```suggestion
By default, Kubernetes recursively changes ownership and permissions for the contents of each
volume to match the `fsGroup` specified in a Pod's `securityContext` when that volume is
mounted.
For large volumes, checking and changing ownership and permissions can take a lot of time,
slowing Pod startup. You can use the `fsGroupChangePolicy` field inside a `securityContext`
to control the way that Kubernetes checks and manages ownership and permissions
for a volume.
```
```

> **fsGroupChangePolicy** -  fsGroupChangePolicy defines behavior of changing ownership and permission of the volume

Reviewer 2: another nit:
```suggestion
**fsGroupChangePolicy** -  `fsGroupChangePolicy` defines behavior for changing ownership and permission of the volume
```

before being exposed inside a Pod. This field only applies to volume types that support
`fsGroup` controlled ownership (and permissions).
This field has no effect on ephemeral volume types such as
[`secret`](https://kubernetes.io/docs/concepts/storage/volumes/#secret),
[`configMap`](https://kubernetes.io/docs/concepts/storage/volumes/#configmap),

> and [`emptydir`](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir). This field has two possible values:

Reviewer 2: Optional change: the detail about ephemeral volumes could go inside a `{{< note >}}` shortcode.

Author: done

* OnRootMismatch : Only change permissions and ownership if permission and ownership of root directory does not match with expected permissions of the volume. This could help shorten the time it takes to change ownership and permission of a volume.

> * Always: Always change permission and ownership of the volume when volume is mounted.

Reviewer 2: nits (only change these if something else also needs fixing):
```suggestion
* _OnRootMismatch_: Only change permissions and ownership if permission and ownership of root directory does not match with expected permissions of the volume. This could help shorten the time it takes to change ownership and permission of a volume.
* _Always_: Always change permission and ownership of the volume when volume is mounted.
```

volume to match the `fsGroup` specified in a Pod's `securityContext` when that volume is
mounted.
For large volumes, checking and changing ownership and permissions can take a lot of time,
slowing Pod startup. You can use the `fsGroupChangePolicy` field inside a `securityContext`
to control the way that Kubernetes checks and manages ownership and permissions
for a volume.

**fsGroupChangePolicy** -  `fsGroupChangePolicy` defines behavior for changing ownership and permission of the volume
before being exposed inside a Pod. This field only applies to volume types that support
`fsGroup` controlled ownership and permissions. This field has two possible values:

* _OnRootMismatch_: Only change permissions and ownership if permission and ownership of root directory does not match with expected permissions of the volume. This could help shorten the time it takes to change ownership and permission of a volume.
* _Always_: Always change permission and ownership of the volume when volume is mounted.

For example:

```yaml
securityContext:
  runAsUser: 1000
  runAsGroup: 3000
  fsGroup: 2000
  fsGroupChangePolicy: "OnRootMismatch"
```

> This is an alpha feature. To use it, enable the [feature gate](/docs/reference/command-line-tools-reference/feature-gates) `ConfigurableFSGroupPolicy` in the kubelet, api-server and controller-manager.

Reviewer 2: nit (should not block release):
```suggestion
This is an alpha feature. To use it, enable the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) `ConfigurableFSGroupPolicy` for the kube-api-server, the kube-controller-manager, and for the kubelet.
```
